sbom-utility-scripts/scripts/merge-cachi2-sboms-script/merge_cachi2_sboms.py

```diff
@@ -148,6 +299,199 @@ def _merge_tools_metadata(syft_sbom: dict[Any, Any], cachi2_sbom: dict[Any, Any]
+        annotation_tuples.append(
+            (
+                annotation["annotator"],
+                annotation["comment"],
+                annotation["annotationDate"],
+                annotation["annotationType"],
+            )
+        )
+    annotations = set(annotation_tuples)
+    return [
+        {
+            "annotator": annotation[0],
+            "comment": annotation[1],
+            "annotationDate": annotation[2],
+            "annotationType": annotation[3],
+        }
+        for annotation in annotations
+    ]
+
+
+def merge_relationships(relationships1, relationships2, packages):
+    """Merge SPDX relationships."""
+
+    def map_relationships(relationships):
+        """Map relationships of spdx element.
+        Method returns triplet containing root element, map of relations and inverse map of relations.
+        Root element is considered as element which is not listed as related document
+        in any of the relationships. Relationship map is dict of {key: value} where key is spdx
+        element and list of related elements is the value.
+        Inverse map is dict of {key: value} where key is related spdx element in the relation ship
+        and value is spdx element.
+        """
+        relations_map = {}
+        relations_inverse_map = {}
+
+        for relation in relationships:
+            relations_map.setdefault(relation["spdxElementId"], []).append(relation["relatedSpdxElement"])
+            relations_inverse_map[relation["relatedSpdxElement"]] = relation["spdxElementId"]
+
+        for parent_element in relations_map.keys():
+            if parent_element not in relations_inverse_map:
+                break
+        return parent_element, relations_map, relations_inverse_map
+
+    def calculate_root_package(root_element, map, inverse_map):
+        """Calculate root package from relationship map.
+        Root package is considered as package which contains other packages and
+        is described by the document itself.
+        """
+        root_package = None
+        for r, contains in map.items():
+            if contains and inverse_map.get(r) == root_element:
+                root_package = r
+        return root_package
+
+    relationships = []
+
+    root_element1, map1, inverse_map1 = map_relationships(relationships1)
+    root_element2, map2, inverse_map2 = map_relationships(relationships2)
+    package_ids = [package["SPDXID"] for package in packages]
+
+    root_package1 = calculate_root_package(root_element1, map1, inverse_map1)
+    root_package2 = calculate_root_package(root_element2, map2, inverse_map2)
+
+    for relation in relationships2:
+        _relation = relation.copy()
+
+        # If relations is Root decribes middle element, skip it
+        if (
+            _relation["relatedSpdxElement"] == root_package2
+            and _relation["spdxElementId"] == root_element2
+            and _relation["relationshipType"] == "DESCRIBES"
```

Comment:
I tried giving the variables less confusing names and the code started to make a bit more sense

```suggestion
            _relation["relatedSpdxElement"] == root2
            and _relation["spdxElementId"] == doc2
            and _relation["relationshipType"] == "DESCRIBES"
```

(inspired by https://github.com/midnightercz/konflux-ci-architecture/blob/spdx-support/ADR/0044-spdx-support.md#relationships)